ni/panels/v1/panel_service.proto

```diff
@@ -0,0 +1,146 @@
+syntax = "proto3";
+
+package ni.panels.v1;
+
+import "google/protobuf/any.proto";
+
+option cc_enable_arenas = true;
+option csharp_namespace = "NationalInstruments.Panels.V1";
+option go_package = "panelsv1";
+option java_multiple_files = true;
+option java_outer_classname = "PanelsProto";
+option java_package = "com.ni.panels.v1";
+option objc_class_prefix = "NIPS";
+option php_namespace = "NI\\Panels\\V1";
+option ruby_package = "NI::Panels::V1";
+
+// Service interface for interacting with NI panels
+service PanelService {
+  // Start a panel using the provided configuration (or connect to if it has already been started)
+  // Status Codes for errors:
+  // - INVALID_ARGUMENT: 
+  //   - The panel identifier contains invalid characters. Only alphanumeric characters and underscores are allowed.
+  //   - The panel configuration has an invalid argument.
+  // - NOT_FOUND: 
+  //   - The panel configuration includes a file that was not found.
+  rpc StartPanel(StartPanelRequest) returns (StartPanelResponse);
+
+  // Stop a panel
+  // Status Codes for errors:
+  // - INVALID_ARGUMENT:
+  //   - The panel identifier contains invalid characters. Only alphanumeric characters and underscores are allowed.
+  rpc StopPanel(StopPanelRequest) returns (StopPanelResponse);
+
+  // Enumerate the panels available in the system, including information about the state of the panels and what values they have.
+  // Status Codes for errors:
+  rpc EnumeratePanels(EnumeratePanelsRequest) returns (EnumeratePanelsResponse);
+
+  // Get a value for a control on the panel
+  // Status Codes for errors:
+  // - INVALID_ARGUMENT: 
+  //   - The panel identifier contains invalid characters. Only alphanumeric characters and underscores are allowed.
+  //   - The value identifier contains invalid characters. Only alphanumeric characters and underscores are allowed.
+  // - NOT_FOUND: 
+  //   - The value with the specified identifier was not found.
+  rpc GetValue(GetValueRequest) returns (GetValueResponse);
+
+  // Try to get a value for a control on the panel
+  // Status Codes for errors:
+  // - INVALID_ARGUMENT: 
+  //   - The panel identifier contains invalid characters. Only alphanumeric characters and underscores are allowed.
+  //   - The value identifier contains invalid characters. Only alphanumeric characters and underscores are allowed.
+  rpc TryGetValue(TryGetValueRequest) returns (TryGetValueResponse);
```

Comment:
What is the difference between this and `GetValue`? I guess this one doesn't return an error if neither the `panel_id` or `value_id` exist or if no value has ever been written to that `value_id`? I guess writing a value is what creates the `value_id` to begin with so maybe there is no notion of a `value_id` existing without a value ever having been written to it?

Reply:
That is correct. If a user tries to get a value before setting a value, then there will be nothing to return. The `TryGetValue` method returns a response without a value whereas the `GetValue` method will error if the value is not set.